### ISSUE TITLE
Push support

### ISSUE THREAD

**Issue by rjahn** [sibvisions/flutter_jvx#152]
Tests required. What's necessary and how could we trigger push notifications from another application/service...

**Comment by Bungeefan** [comment]
## Summary:

#### We need to use the platform-specific Push platform.
Means Android - [FCM](https://firebase.google.com/docs/cloud-messaging) and iOS - [APNS](https://developer.apple.com/documentation/usernotifications).
In the case of iOS, there simply is no way around and for Android we **theoretically** could make a custom background service, but that would probably end in a much worse user experience and higher maintenance.
Besides that, both services are free-of-charge even in case of Firebase (yes also for commercial).

### How it works:

#### Mobile device:

The mobile device registers with their associated notification service and receives a token in return.
This token, also known as device token (APNS) or registration token (FCM), has to be sent to our application server.

The token can and will change over time, therefore we should probably send it to the application server on startup and when a new token was assigned (we receive an event for this).

#### Application server (e.g. JVx.mobile):

The application server then has to save the received device tokens and associate them with the correct client. These tokens are required to specify the target device for a notification.
In short, to deliver a notification, a POST request is sent to FCM/APNS with a title and a message as well as an optional data map. The request also has to contain the device token (to identify the receiving mobile device) and the signed JWT, also known as provider token.

##### Authentication (signed JWT)

Apple only supports JWT-based authentication to their APNS and FCM supports JWT too along with multiple other authentication methods.
Fortunately, we can use a similar JWT for both, the main difference is in the signing of the token.
Apple provides us with an ECDSA private key and FCM with an RSA key.

### A note for custom apps:

The setup procedure has to be done for each compiled app and these apps must be registered in a developer account which in turn shares the same private key for all their apps. Therefore, each custom app user of our framework would need some native setup and a separate developer account for every platform.

In simple terms, the one who hosts the application server should also own the associated developer account, as the server needs the keys to function.

#### Relay service

For users who want us to deploy their app in the stores but host the application server themselves, we could create a relay service, therefore, their server could send a notification request to our service which in turn uses our private key to contact APNS/FCM.

### Useful links:

* APNS
  * https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/establishing_a_token-based_connection_to_apns
  * https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/generating_a_remote_notification
* FCM
  * https://developers.google.com/identity/protocols/oauth2/service-account#jwt-auth
  * https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#Message

**Comment by Bungeefan** [comment]
1be9b727693ae70caccfb9290d8da616f1f6525d

**Comment by Bungeefan** [comment]
There is also a promising looking PR for the vanishing local notifications issue:
https://github.com/ben-xD/push/pull/33

Can be tested with:
```yaml
dependency_overrides:
  push_ios:
    git:
      url: https://github.com/Arley011/push
      path: push_ios
      ref: local-notification-present-parameters
```

**Comment by Bungeefan** [comment]
Update:
Local notifications issue has been resolved with push 1.0.5 :tada:
